pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,370 @@
+			glog.Errorf("Unable to remove directory: %s", err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
+// writeDataInTimestampDir writes the latest data into a new temporary directory with a timestamp.
+func (d *downwardAPIVolume) writeDataInTimestampDir(data map[string]string) (string, error) {
+	errlist := []error{}
+	timestampDir, err := ioutil.TempDir(d.GetPath(), ".."+time.Now().Format("2006_01_02_15_04_05"))
+	for fileName, values := range data {
+		fullPathFile := path.Join(timestampDir, fileName)
+		dir, _ := filepath.Split(fullPathFile)
+		if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+			glog.Errorf("Unable to create directory `%s`: %s", dir, err.Error())
+			return "", err
+		}
+		if err := ioutil.WriteFile(fullPathFile, []byte(values), 0644); err != nil {
+			glog.Errorf("Unable to write file `%s`: %s", fullPathFile, err.Error())
+			errlist = append(errlist, err)
+		}
+	}
+	return timestampDir, utilErrors.NewAggregate(errlist)
+}
+
+// updateSymlinksToCurrentDir creates the relative symlinks for all the files configured in this volume.
+// If the directory in a file path does not exist, it is created.
+//
+// For example for files: "bar", "foo/bar", "baz/bar", "foo/baz/blah"
+// the following symlinks and subdirectory are created:
+// bar          -> ..downwardapi/bar
+// baz/bar      -> ../..downwardapi/baz/bar
+// foo/bar      -> ../..downwardapi/foo/bar
+// foo/baz/blah -> ../../..downwardapi/foo/baz/blah
+func (d *downwardAPIVolume) updateSymlinksToCurrentDir() error {
+	for _, f := range d.fieldReferenceFileNames {
+		dir, _ := filepath.Split(f)
+		nbOfSubdir := 0
+		if len(dir) > 0 {
+			// if dir is not empty f contains at least a subdirectory (for example: f="foo/bar")
+			// since filepath.Split leaves a trailing '/'  we have dir="foo/"
+			// and since len(strings.Split"foo/")=2 to count the number
+			// of sub directory you need to remove 1
+			nbOfSubdir = len(strings.Split(dir, "/")) - 1
+			if err := os.MkdirAll(path.Join(d.GetPath(), dir), os.ModePerm); err != nil {
+				return err
+			}
+		}
+		if _, err := os.Readlink(path.Join(d.GetPath(), f)); err != nil {
+			// link does not exist create it
+			if err := os.Symlink(path.Join(strings.Repeat("../", nbOfSubdir), downwardAPIDir, f), path.Join(d.GetPath(), f)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// readFile reads the file at the given path and returns the content as a string.
+func readFile(path string) string {
+	if data, err := ioutil.ReadFile(path); err == nil {
+		return string(data)
+	}
+	return ""
+}
+
+// sortLines sorts the strings generated from map based data
+// (annotations and labels)
+func sortLines(values string) string {
+	splitted := strings.Split(values, "\n")
+	sort.Strings(splitted)
+	return strings.Join(splitted, "\n")
+}
+
+func (d *downwardAPIVolume) GetPath() string {
+	return d.plugin.host.GetPodVolumeDir(d.pod.UID, util.EscapeQualifiedNameForDisk(downwardAPIPluginName), d.volName)
```

Comment:
What I have done is just made a podUID field in the shared struct so that GetPath() can be used from the cleaner too.